## Tutorial 2

**Objective:** To perform reference-based assemblies of SARS-CoV-2 genomes to determine whether cases from patients and staff at a facility are part of the same
transmission chain.

You will map reads to a reference assembly, generate consensus genomes, identify single nucleotide variants (SNVs),
and phylogenetically place genomes to determine whether the cases in question are epidemiologically related.
Additionally, we will employ and compare the results of different methods to generate our consensus genomes.
This includes using software that I have pre-installed on the VM as well as containerized versions of the same
software pulled from Docker. Finally, you will learn about GitHub and Docker for obtaining and installing bioinforamtics software.
<br>

## Some general information

Our read data are in fastq format. Fastq files contain the quality score for each nucleotide in the read and each read has four lines associated with it:

	>@SRR10971381.1 1 length=151
	AGTCGATCAGCTGAGTACTAGTAGCATG
	+
	BBBCCCKKIBCCKKBIJDFFHHHIIJKK

> The read name, naming conventions may vary but names are always prefaced by '@'<br>
> This is followed by the sequence line <br>
> `+` a Line break <br>
> Followed by quality scores for each nucleotide in ASCII format

<br>

Each base has an associated quality score, which indicates the reliability of that call.
Phred scores are equal to -10 log<sub>10</sub>P, where P is the error probability for the base in question.
Thus, a phred score of 10 is equal to an error probability of 0.01 (P=10<sup>â€“Q/10</sup>) or a 1 in 10 chance that the base call is incorrect.

Most NGS output is in fastq format.  We are working with short, paired-end reads (2x250 bp or 2x150 bp)
generated by Illumina sequencers (in this case a MiSeq or a NextSeq).
While Illumina reads are shorter than those from traditional Sanger sequencing, Illumina sequencing can produce millions to hundreds of millions of reads per sample.
<br>


## General workflow for producing a reference-based assembly

Reference-based assemblies rely on a reference genome on which to map the reads of a genomic library.
Reference assemblies allow the direct comparison of a genome of interest to an already-assembled genome for SNP and/or Indel detection.
Not all differences detected will be legitimate variation.  Miss-assembly (usually due to repetitive elements),
poor quality reads, low coverage, and contamination can all contribute to erroneous base calls.
Implementing additional filtering criteria to detect true variation is necessary,
such as only considering SNPs from an area of high coverage with a certain number of high-quality reads.
A general workflow for generating a reference-based assembly is given below:

<br>

**Quality control**

  - Generate some summary statistics with FastQC

**Adaptor removal (and primer removal) and quality trimming**

 - cutadapt
 - Trimmomatic
 - BBduk from BBtools
 - FastX-toolkit
 - TrimGalore
 - Seqyclean
 - And many more

**Read correction (optional, this is more important for long-read sequencing technologies such as Nanopore)**

 - Quake
 - BBnorm
 - SHRIMP
 - Quorum
 - Lighter
 - Fermi
 - Musket
 - And many, many more

**Read alignment or mapping to reference genome**

 - BWA
 - Bowtie
 - BBmap
 - STAR (for RNA-Seq)

**General statistics pertaining to reference assembly**

 - Samtools
 - Picard
 - BBtools
 - Quast

**SNP and INDEL calling**

 - Samtools/bcftools
 - GATK
 - iVAR

**Additional SNV filtering and annotation**

In our case, we'll also be generating consensus genomes with iVAR

<br>

## Background information for SARS-CoV-2 cases under investigation
<br>
In early March 2020 of the pandemic, healthcare facilities struggled with outbreaks of COVID-19.
Genomic epidemiology can assist investigations into the source of an outbreak, the timing of introduction
of a pathogen into a population, and the number of introductions that have occurred into an area.
Here we are using the information contained within SARS-CoV-2 genomes to assist a traditional epidemiological investigation
into the nature of an outbreak in a Westchester healthcare facility. The are four samples from the facility: two
from staff and two from residents, collected on the same day. A fifth sample is also from a healthcare facility collected four days after.
While genomic epidemiology cannot confirm the individual source
in this case, we want to know if the virus was transmitted within the facility and if all cases are a part of the same outbreak.
Keep in mind that the mutation rate for SARS-CoV-2 is about 1 mutation every two weeks.  At the beginning of the pandemic,
there was very little variation in SARS-CoV-2 genomes.  Thus, one unique mutation shared among several genomes that is not observed
in a contextual set of genomes might provide support for a transmission link.  Conversely, it could be that there are many unsampled
genomes, some of which also contain the unique mutation.

<br>

Each student will map the fastqs from one sample to the reference genome and generate a consensus genome for downstream analyses.

## Sample assignment and metadata

| Student | Collection date | Sample name | County | Healthcare facility |
| ------- | --------------- | ----------- | ------ | ------------------- |
| Kathleen McCarthy | 2020-03-15 | IDR2000016084 | Westchester | Yes, staff |
| Kayla Simanek | 2020-03-15 | IDR2000016115 | Westchester | Yes, staff |
| Nicholas Keegan | 2020-03-15 | IDR2000016158 | Westchester | Yes, resident |
| Rachel Lange | 2020-03-15 | IDR2000016600 | Westchester | Yes, resident |
| Sharon Shaughnessy | 2020-03-19 | IDR2000024179 | Westchester | Yes, resident |



## Follow the general workflow outlined previously to process and map your reads and generate a consensus genome

<br>

Use the **`gsutil`** copy command to copy the fastq files for your sample from our bucket to the terminal.

Before we can copy to and from our bucket, we need to authenticate our accounts to GCP on the computer from which we are executing the gsutil commands.

	gcloud auth login

> A crazy long URL is output. Copy the URL from the terminal window and paste it into a browser (preferably Chrome).
> The web page that is displayed will contain a long auth code. Copy the code and paste it back into the terminal window
> you executed the gcloud cmd in. press return and you should be authenticated to GCP. Some current state info will be output.

<br>

	gsutil cp gs://wc-bms-bi-training-bucket/outbreak_fastqs/IDR2000024179*gz .
	
> Make sure you substitute your IDR number in the command above. The ending period is important! Remember, you must specify a destination for your copied files, 
> which is you current working directory in this case.

<br>

Clean your raw reads and remove any remaining adapters with TrimGalore, where fastq_R1 and fastq_R2 are your fastq files.

	trim_galore -q 20 --length 100 --paired fastq_R1 fastq_R2
	
> Adapters are short oligonucleotides ligated to a library for sequencing on Illumina machines. These are typically removed by the Sequencing Core but some 
> may still remain. We might also want to remove reads that fall below a certain average quality or length.
> Before running trim_galore, explore the options available to you with the help command. Most default setting are fine.
>
> Here we are using the default quality threshold of 20 (so we don't even have to specify it) and a minimum sequence length of 100. 
> **`Paired`** keeps R1 and R2 reads in order.  In other words, if one of the pairs fails quality control, both are removed.  
> The nice thing about TrimGalore is it pretty much does everything for you automatically, including detecting the type of adapter present 
> It also prints some nice summary statistics to STDOUT.

* Judging from the TrimGalore output, does your sequencing run look good?
* What is the percentage of reads that passed trimming?
* You can also look at the quality with [FastQC](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/)

<br>

Create an index of your reference genome.

Your adapter, quality trimmed reads are now in files with a "val_[1,2].fq.gz" ending. We'll map these to the Wuhan-1 reference genome.
Before we perform the read alignment, we need to index the reference assembly. You might want to rename your fna file to something shorter for easier
viewing of long commands in your terminal window.  I've renamed mine "wuhan.fna."
