<HTML><head><title>Bioinformatics Workshop</title></head><body>
<style>
#h3 {
 background: lightgray;
 white-space: nowrap;
}

table, th, td {
  border:1px solid black;
}

ul.a {
  list-style-type: circle;
}

#metadata {
	border-collapse:collapse;
	width:100%;
}

#metadata td, #metadata th {
	border: 1px solid #ddd;
	padding: 8px;}
}

#metadata tr:nth-child(even){background-color:#f2f2f2;}
#metadata tr:hover {background-color:#ddd;}

#metadata th {
	padding-top: 12px;
	padding-bottom: 12px;
	text-align:left;
	background-color: green;
	color: white;
}

pre.code {
        background-color: #EEE;
        padding: 5px 0px;
        margin-top: 5px;
        margin-bottom: 15px;
}

</style>

<h1>Tutorial 3</h1>

<p><b>Objective: </b>To perform a <i>de novo</i> assembly of the Wuhan-1 genome using two different assemblers and compare the results.</p>
<br>

<p>Regardless of the sequencing technology (Sanger, 454, Illumina, PacBio, Nanopore, etc.)
and the assembly algorithm (de bruijn graph, overlap layout consensus, etc.),
all <i>de novo</i> genome assembly programs essentially rely on stringing together overlapping
portions of reads into longer fragments called contigs (contiguous sequences).
In turn, these contigs can be joined to generate scaffolds if you have paired-end reads (forward and reverse reads). Here
we will attempt to recover the original Wuhan-1 genome, which was helped to identify SARS-CoV-2 as a novel Betacoronavirus.
SPAdes and MEGAHIT are two of the most frequently employed assemblers for bacterial/viral genomes.  MEGAHIT was designed for metagenomes
but can also work on single isolate libraries. Shovill is basically a faster adaptation of SPAdes.</p>
<br>

<hr>
<p><b>Trim adapters and quality filter<p></b>
<hr>
</li>Use TrimGalore (or try BBtools!) to filter your reads.</li>
<blockquote>
<pre class="code">
trim_galore -q 20 --paired SRR10971381_1.fastq SRR10971381_2.fastq
</pre>
<p>We would hope that reads submitted to the SRA have adapters trimmed, but this is often not the case.</p>
</blockquote>
</br>
<li>Rename your files something less clunky.</li>
<blockquote>
<pre class="code">
mv SRR10971381_1_val_1.fq SRR10971381-trimmed_1.fq
mv SRR10971381_2_val_2.fq SRR10971381-trimmed_2.fq
</pre>
</blockquote>

<hr>
<p><b>Downsample your trimmed fastq files</b></p>
<hr>
<li>Use the <code>sample</code> function of seqtk to randomly subsample your fastq files.</li>
<blockquote>
<pre class="code">
seqtk sample -s100 SRR10971381-trimmed_1.fq 5000000 > SRR10971381-sub_1.fastq
seqtk sample -s100 SRR10971381-trimmed_2.fq 5000000 > SRR10971381-sub_2.fastq
</pre>
<p><a style='color:blue' href="https://github.com/lh3/seqtk">Setqk</a> is like a swiss-army knife for manipulating fasta/fastq files.
Two ensure that we randomly subsample the same reads from both R1 and R2 files (maintain the files as paired),
we need to use the same starting seed (specified by the '-s' argument). Subsampling is necessary because our fastq files are over 8 Gb in size (>28 million reads)!
We don't have enough memory on our VMs to run a full assembly.</p>

<p>Since this was a metatranscriptome from a human subject, we might expect the majority of our reads to be human-derived. We
might want to exclude these sequences as well before assembling to decrease the run-time.  In this case, because we downloaded the fastq files
from the SRA, all human DNA/RNA should be removed already.</p>
</blockquote>
<br>

<li>Copy a slightly modified, downsampled version of the original fastqs from our bucket to your VM.</li>
<blockquote>
<pre class="code">
gsutil -m cp gs://wc-bms-bi-training-bucket/reference_assembly/downsampled*fastq .
</pre>
<p>This is a further downsampled version of the original data.  Use these fastqs for all downstream processes in this section of the tutorial.</p>
</blockquote>
<br>

<hr>
<p><b>Perform a metagenome assembly with SPAdes</b></p>
<hr>

<li>Assemble the metagenomic reads from our SRA download in Tutorial 1 with <a style='color:blue' href="https://github.com/ablab/spades">SPAdes</a>.</li>
<blockquote>
<pre class="code">
docker run --rm -v $(pwd):/data -w /data staphb/spades spades -1 downsampled-trimmed_1.fastq -2 downsampled-trimmed_2.fastq -o wuhan_assembly_spades/ --meta &
</pre>
<p>SPAdes provides several assembly options (including the metagenome option) and is ideally suited for smaller genomes.
SPAdes will perform error-correction prior to assembling or you can perform error-correction independently.

<p><b>Note:</b> The ampersand specifies performing a job in the background. This returns your command-line prompt so that you can continue working.
Given the size of the library (8 GB!), assembly could take quite some time if we used all reads. You can see that your job is stil running by typing <code>top</code>,
which allows you to see the status of all jobs on your VM. But because SPAdes prints a lot of information to STDOUT, it's rather inconvenient to work
in this terminal.  You can start a new connection by selecting this option in the drop down menu that looks like a cog in the upper right corner of your terminal.</p>
</blockquote>
<br>

<li>Generate some summary statistics for your assembly</li>
<blockquote>
<p>The final assembly is located in the 'scaffolds.fasta' file in the output directory you specified.
Ideally, we would like a closed genome (one chromosome).
Although short reads are typically not enough to scaffold contigs into chromosome-sized pieces, our genome is only
29903 base pairs long.</p>
<ol>
<li value='1'>
How many contigs are in the “scaffolds.fasta” file?</li>
<li>What is the average length of your scaffolds?  Hint: the length of each scaffold is given in the definition line.</li>
<li>What is the N50 of your assembly (N50 is the value whereby 50% of an assembly is contained in contigs/scaffolds equal to or larger than this value)?</li>
<li>What is the average coverage of your scaffolds? Hint: the average coverage of each scaffold is also given in the definition line.</li>
<li>Since there are several contigs in this file, how can we identify which are SARS-CoV-2 (Hint: try BLAST)?</li>
<li>If there are several SARS-CoV-2 contigs, how can we identify the Wuhan-1 genome?</li>
</ol>
</blockquote>
<br>

<hr>
<p><b>BLAST (Basic Local Alignment Search Tool)</b></p>
<hr>
<p><a style='color:blue' href="https://blast.ncbi.nlm.nih.gov/Blast.cgi">BLAST</a> looks for regions of similarity between sequences, which serves as a proxy of homology (descended from a common ancestor).
We could use the website to perform a BLAST search and identify similar sequences to our assembled contigs our we could use a locally installed version.  Similarly, we could use
the nucleotide database provided by NCBI (all nucleotide sequences deposited at NCBI) or we could make our own database. Since we are only trying to identify the Wuhan-1 SARS-CoV-2 genome
from our assembly, we could simply make a database of the Wuhan-1 reference genome and BLAST to it.</p>
<li>Make a BLAST database of the Wuhan-1 genome downloaded from NCBI in Tutorial 1</li>
<blockquote>
<pre class="code">
makeblastdb -in wuhan.fna -dbtype nucl
</pre>
<p> Here we are specifying the input file as well as the database type, which is nucleotide.  We could also generate a database from
protein sequences, which would take the argument 'dbtype prot.'
</blockquote>
<br>
<li>Blast your assembly to the Wuhan-1 genome.</li>
<blockquote>
<pre class="code">
blastn -query wuhan_assembly_spades/scaffolds.fasta -db wuhan.fna -outfmt 6 -out spades.br
</pre>
<p>There are different flavors of blast depending on the query and database type (nucleotide vs. protein).
The <code>blastn</code>' command specifies a nucleotide to nucleotide search.  We can modify the format of
the output with the <code>-outfmt</code> option.  Here we are saving our output to a tab-delimited file called 'spades.br.'</p>

<pre style='font-family:Times New Roman'>
Did you recover the Wuhan-1 genome?
How similar is it to the reference?
What do all of the fields mean (use <code>blastn -help</code> to see the full list of blast options)?
</pre>


<br>
<br>
</body>
</HTML>