<HTML><head><title>Bioinformatics Workshop</title></head><body>
<style>
#h1 {
 background: lightgray;
 display: inline-block;
 position: relative;
 white-space: nowrap;
 <!--width:0.9%;-->
}

#h2 {
 background: lightgray;
 display: inline-block;
 position: relative;
 left: 4%;
 transform: translateX(-50%);
 white-space: nowrap;
 <!--width:5.2%;-->
}

#h3 {
 background: lightgray;
 white-space: nowrap;
}


table, th, td {
  border:1px solid black;
}

pre.code {
        background-color: #EEE;
        padding: 5px 0px;
        margin-top: 5px;
        margin-bottom: 15px;
}

</style>

<h1>Tutorial 1 continued</h1>
<br>

<hr>
<p><b>Databases and obtaining sequences</b></p>
<hr>

<p>There are several sequence databases – NCBI, JGI, EMBL - that you might encounter.
You might want to explore each of these to familiarize yourself with the resources they offer.
We will focus on NCBI.  Our goal is to download a reference genome (Wuhan-1) for SARS-CoV-2,
which we will use later to perform a reference-based assembly and an outbreak analysis.
We will also want to download the accompanying annotation file (gff file), which provides a description of the genes,
the function of the coding sequences, and the nucleotide positions of the genes in the genome and the translated coding sequences (faa file).</p>
<br>

<li> Navigate to NCBI’s homepage: <a style='color:blue' href='https://www.ncbi.nlm.nih.gov/'>https://www.ncbi.nlm.nih.gov/</a></li>
<blockquote>
<p> Notice that there are options to submit sequences, download sequences, and even analyze data.
PubMed allows literature searches and BLAST is an alignment tool to look for sequences that are similar (a proxy for homology) to your queries.
Also notice that NCBI has provided a quick link to SARS-CoV-2 data. You could obtain a nucleotide record by clinking on this link but we'll follow
the more traditional route for now.</p></blockquote>
<br>

<li> Choose “Genome” under the top left pull-down menu (set to “All Databases” by default), type SARS-CoV-2 into the search area, and hit enter.</li>
<blockquote>This brings us to a page containing information on the reference genome (Wuhan-1).  We could also 'Browse the list' of other available genomes.
We can click on the 'RefSeq' link and use the 'Send to' menu to save the 'Complete Record' for this genome to a file in 'fasta' format. By default,
the file is named 'sequence.txt.'  From there we can upload the file to our VM. Instead we will:</blockquote>
<br>

<li>'Browse the list' of available genomes to access the RefSeq or GenBank FTP site for our reference genome (which is conveniently the first genome listed).</li>
<blockquote> If it wasn't the first genome listed, we could apply search filters to narrow the list.</blockquote>
<br>

<li> Click on the RefSeq ('R') FTP link for MN908497.3 (Wuhan-1). Open the link as a "Guest."</li>
<blockquote> This takes you to a directory with several files associated with the Wuhan-1 genome. For now let's obtain the genome assembly (.fna), the accompanying annotation (.gff), and the protein coding sequences (.faa).
Instead of a two-step process of downloading files to our computers and then uploading them to the cloud, we will use the 'curl' command (copy url) to copy the files directly to
our VMs. </blockquote>
<br>

<li> Right click on the fna file and 'Get info.'  </li>
<blockquote> The information page will list the file location on NCBI's server. </blockquote>
<br>

<li> Copy the server information and type the following command in your VM terminal:</li>
<blockquote>
<pre class="code">
<span style='color:blue'>curl -O</span> <span style='color:red'>"ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/009/858/895/GCF_009858895.2_ASM985889v3/GCF_009858895.2_ASM985889v3_genomic.fna.gz"</span>
</pre>
<p>The -O option saves the copied contents to a file named as it was on the FTP site.</p>
</blockquote>
<li> Repeat this for the gff and faa files </li>
<br>

<hr>
<p><b>Obtaining reads from the SRA</b></p>
<hr>

<p> Now let's download the raw reads for the Wuhan-1 reference genome from the SRA (sequence read archive) database.  We'll use these later to perform a de-novo assembly.
Typically, any published NGS data must also be submitted to the SRA. Each sample/specimen sequenced will have a BioSample accession number. Biosample information provides associated
metadata.  The SRA and Biosample for each submission are further housed under a BioProject, which can contain multiple submissions from the same study or experiment</p>
<br>

<li>Return to the <a style='color:blue' href='https://www.ncbi.nlm.nih.gov/'>NCBI homepage</a> and search genomes for Wuhan-1 again. This brings you to its GenBank page.</li>
<blockquote>Normally, we would click the link to the associated BioProject but this takes us to a large umbrella project that would be difficult to search.
Instead, we could click the link to the publication on PubMed and search the journal article for the BioProject.  I have done this for you.</blockquote>
<br

<li>Go to the Bioproject for Wuhan 1: <a style='color:blue' href='https://www.ncbi.nlm.nih.gov/bioproject/?term=PRJNA603194'>PRJNA603194</a>
<blockquote>Notice the BioProject page gives you information on the purpose of the study, as well as several links, including one to the SRA experiment.</blockquote>
<br>

<li>Click on the link to the <a style='color:blue' href='https://www.ncbi.nlm.nih.gov/sra?linkname=bioproject_sra_all&from_uid=603194'>SRA experiment</a> for Wuhan-1</li>
<blockquote>Notice this is a metatranscriptome, meaning this may not be a pure isolate of a single SARS-CoV-2 virus (although human RNA/DNA should have been removed prior to
SRA submission).  The sequencing was performed on an Illumina MiniSeq and the reads are paired (i.e. there are Forward and Reverse reads for the same amplicon).
We could download the two fastq files using the NCBI link provided or we could use faster tools provided by NCBI.</blockquote>
<br>

<li>Use 'prefetch' and 'fasterq-dump' tools from the SRA toolkit to download the Wuhan-1 fastq files</li>
<blockquote>
<pre class="code">
<span style='color:blue'>prefetch </span> <span style='color:red'>SRR10971381</span>
<span style='color:blue'>fasterq-dump </span> <span style='color:red'>SRR10971381</span>
</pre>
<p>Prefetch will download the SRA data in binary format and fasterq-dump will perform the fastq conversion.
Notice that the conversion tool automatically saves forward ("_1.fastq") and reverse reads ("_2.fastq"") to separate files.
Although we could use fasterq by itself, NCBI claims prefetch in combination with fasterq is faster.
More information on the SRA-toolkit and other frequently used tools can be found here: <a style='color:blue' href="https://trace.ncbi.nlm.nih.gov/Traces/sra/sra.cgi?view=toolkit_doc">SRA toolkit</a></p>
</blockquote>
<br>

<p><b>Note:</b> If your are having trouble obtaining the fna, faa, and gff files from NCBI, you can also copy them to your account on the VM from our GCP bucket using
the gsutil commands provided by GCP. Fastq files also available in the bucket in the same folder: 'reference_assembly.' However, you will need to authenticate
your account first in order to use the gsutil commands.</p>
<blockquote>
<pre class="code">
<span style='color:blue'>gsutil -m cp</span> <span style='color:red'>gs://wc-bms-bi-training-bucket/reference_assembly/GCF* .</span>
</pre>
We'll discuss this command more in class but essentially, we are copying (cp) all files that start with GCF from our GCP bucket to our current directory (.).
</blockquote>
<br>
<br>
<br>

</body>
</HTML>