<HTML><head><title>Bioinformatics Workshop</title></head><body>
<style>
#h3 {
 background: lightgray;
 white-space: nowrap;
}

table, th, td {
  border:1px solid black;
}

ul.a {
  list-style-type: circle;
}

pre.code {
        background-color: #EEE;
        padding: 5px 0px;
        margin-top: 5px;
        margin-bottom: 15px;
}

</style>

<h1>Tutorial 3 continued</h1>
<br>
<p> Now we will use our full fastq files downloaded from the SRA.</p>
<br>

<hr>
<p><b>Perform the same assembly with Shovill</b></p>
<hr>
<li>Use a Dockerized version of Shovill to assemble the full fastq files from the SRA.</li>
<blockquote>
<pre class="code">
docker run --rm -v $(pwd):/data -w /data staphb/shovill shovill --R1 SRR10971381_1.fastq --r2 SRR10971381_2.fastq --trim --outdir wuhan_assembly_shovill
</pre>
<p>As mentioned previously, Shovill is a faster implementation of SPAdes. Notice how much faster Shovill ran even with all of our data!
We are using the <code>--trim</code> option to remove adapters. Combined with the automatic read-error correction implemented, we can start with
our raw reads.  The one caveat is that Shovill doesn't have a metagenome mode.</p>
</blockquote>
<li>Briefly explore the Shovill assembly and its summary output to see how the program compares to SPAdes.</li>
<blockquote>
<p>The final contigs are in a file called 'contigs.fa' and are conveniently organized by descending length. Also notice that the naming conventions
for the contigs are much more manageable (although the original name assigned by SPAdes is also included).</p>
<pre style='font-family:Times New Roman'>
Did Shovill recover the Wuhan genome?
What assembler would you choose based on this exercise?</pre>
</blockquote>

<hr>
<p><b>Perform a metagenome assembly with Megahit.</p></b>
<hr>

<li>Use default parameters to assemble the metagenome with MEGAHIT.</li>
<blockquote>
<pre class="code">
megahit -1 SRR10971381-trimmed_1.fastq -2 SRR10971381-trimmed_2.fastq -o wuhan_assembly_megahit
</pre>
<p>Notice how much faster MEGAHIT completed in comparison to SPAdes. Your contigs are located in the file called 'final.contigs.fa' in your
output directory <code>wuhan_assembly_megahit</code>.
<ol>
<li value='1'>
Generate the same summary statistics for your MEGAHIT assembly as you did for your SPAdes assembly.</li>
<li>Did all assemblers recover the Wuhan-1 genome? </li>
<li>If so, are they similar lengths and coverage?</li>
<li>Other tools to try are: <a style='color:blue' href="https://pangolin.cog-uk.io/">Pangolin</a>
(also available as a Docker image on your VM) and <a style='color:blue' href='https://clades.nextstrain.org/'>Nextclade</a>.
</ol>
</blockquote>
<br>

<hr>
<p><b>Compare the performance of SPAdes, Shovill, and MEGAHIT with <a style='color:blue' href="https://github.com/ablab/quast">QUAST</a></p></b>
<hr>

<p>Quast is a genome (and metagenome) evaluation tool that can compare the quality of assemblies produced by different programs.
It will output the number of large contigs for each assembly, the length of the longest one, the N50, and the number of genes predicted.
When a reference assembly is provided, Quast will output additional statistics, such as the number of misassemblies, the percentage of the reference genome
recovered, etc.</p>

<li>Copy your contig files from your three assembly directories to the directory with your reference assembly and rename them something meaningful</li>
<blockquote>
<pre class="code">
cp ~/wuhan_fastqs/wuhan_assembly_spades/scaffolds.fasta ~/reference_assembly/spades.fa
cp ~/wuhan_fastqs/wuhan_assembly_shovill/contigs.fa ~/reference_assembly/shovill.fa
cp ~/wuhan_fastqs/wuhan_assembly_megahit ~/reference_assembly/megahit.fa
</pre>
<p> This command assumes your assembly output directories are in a directory called 'wuhan_fastqs' and the reference assembly and annotation file for Wuhan-1
from NCBI is in the directory 'reference_assembly.' The <code>~</code> symbolizes your home directory.</p>
</blockquote>
<br>

<li>Run the Dockerized version of Quast to comare the three assemblies.</li>
<blockquote>
<pre class="code">
cd wuhan_fastqs
docker run --rm -v $(pwd):/data -w /data staphb/quast quast.py -r wuhan.fna -g GCF_009858895.2_ASM985889v3_genomic.gff spades.fa shovill.fa megahit.fa
</pre>
<p>Your results will be located in a directory called 'quast_results' in a subdirectory called 'latest.' Notice the summary report is given to you in multiple formats.</p>
</blockquote>
<br>

<li>Download the html version of the results ('report.html') to your computer.</li>
<blockquote>
<p>The html report provides a nice visual of the results in table and figure formats. Because
we provided a reference genome and a GFF annotation file, Quast compares the performance of
all three assemblers in recovering a complete and accurate (misassemblies, INDELs, etc) Wuhan-1 genome
as well as the number of complete genomic features.</p>
<ol>
<li value='1'>
Which assembler performed best in recovering the Wuhan-1 genome?</li>
<li>What characteristics would you want to see in a metagenome assembly?
<li>Given those characteristics, which created a better metagenome assembly?</li>
</ol>
</blockquote>



<br>
<br>
</body>
</HTML>