<HTML><head><title>Bioinformatics Workshop</title></head><body>
<style>
#h3 {
 background: lightgray;
 white-space: nowrap;
}

table, th, td {
  border:1px solid black;
}

ul.a {
  list-style-type: circle;
}

#metadata {
	border-collapse:collapse;
	width:100%;
}

#metadata td, #metadata th {
	border: 1px solid #ddd;
	padding: 8px;}
}

#metadata tr:nth-child(even){background-color:#f2f2f2;}
#metadata tr:hover {background-color:#ddd;}

#metadata th {
	padding-top: 12px;
	padding-bottom: 12px;
	text-align:left;
	background-color: green;
	color: white;
}

</style>

<h1>Tutorial 2</h1>

<p><b>Objective:</b> To perform reference-based assemblies of SARS-CoV-2 genomes to determine whether cases from patients and staff at a facility are part of the same
transmission chain.</p>

<p>You will map reads to a reference assembly, generate consensus genomes, identify single nucleotide variants (SNVs),
and phylogenetically place genomes to determine whether the cases in question are epidemiologically related.
Additionally, we will employ and compare the results of different methods to generate our consensus genomes.
This includes using software that I have pre-installed on the VM as well as containerized versions of the same
software pulled from Docker. Finally, you will learn about GitHub and Docker for obtaining and installing bioinforamtics software.</p>
<br>

<hr>
<p><b> Some general information</b></p>
<hr>
<p> Our read data are in fastq format. Fastq files contain the quality score for each nucleotide in the read and each read has four lines associated with it:</p>
<br>
<table style='width:50%'>
	<tr>
		<td>@SRR10971381.1 1 length=151</td>
		<td> The read name, naming conventions may vary but names are always prefaced by '@' </td>
	</tr>
	<tr>
		<td>AGTCGATCAG</td>
		<td> The nucleotide sequence</td>
	</tr>
	<tr>
		<td>+</td>
		<td>Line break</td>
	</tr>
	<tr>
		<td>BBBCCCKKI</td>
		<td>Quality scores for each nucleotide in ASCII format</td>
	</tr>
</table>
<br>

<p>Each base has an associated quality score, which indicates the reliability of that call.
Phred scores are equal to -10 log<sub>10</sub>P, where P is the error probability for the base in question.
Thus, a phred score of 10 is equal to an error probability of 0.01 (P=10<sup>â€“Q/10</sup>) or a 1 in 10 chance that the base call is incorrect.</p>

<p>Most NGS output is in fastq format.  We are working with short, paired-end reads (2x250 bp or 2x150 bp)
generated by Illumina sequencers (in this case a MiSeq or a NextSeq).
While Illumina reads are shorter than those from traditional Sanger sequencing, Illumina sequencing can produce millions to hundreds of millions of reads per sample.</p>
<br>

<hr>
<p><b> General workflow for producing a reference-based assembly</b></p>
<hr>

<p>Reference-based assemblies rely on a reference genome on which to map the reads of a genomic library.
Reference assemblies allow the direct comparison of a genome of interest to an already-assembled genome for SNP and/or Indel detection.
Not all differences detected will be legitimate variation.  Miss-assembly (usually due to repetitive elements),
poor quality reads, low coverage, and contamination can all contribute to erroneous base calls.
Implementing additional filtering criteria to detect true variation is necessary,
such as only considering SNPs from an area of high coverage with a certain number of high-quality reads.
A general workflow for generating a reference-based assembly is given below:</p>
<blockquote>
<br>

<b> Quality control </b>
<ul class='a'>
	<li>Generate some summary statistics with FastQC</li>
</ul>
<b> Adaptor removal (and primer removal) and quality trimming</b>
<ul class='a'>
	<li>cutadapt</li>
	<li>Trimmomatic</li>
	<li>BBduk from BBtools</li>
	<li>FastX-toolkit</li>
	<li>TrimGalore</li>
	<li>Seqyclean</li>
	<li>And many more</li>
</ul>
<b> Read correction (optional, this is more important for long-read sequencing technologies such as Nanopore)</b>
<ul class='a'>
	<li>Quake</li>
	<li>BBnorm</li>
	<li>SHRIMP</li>
	<li>Quorum</li>
	<li>Lighter</li>
	<li>Fermi</li>
	<li>Musket</li>
	<li>And many, many more</li>
</ul>
<b> Read alignment or mapping to reference genome</b>
<ul class='a'>
	<li>BWA</li>
	<li>Bowtie</li>
	<li>BBmap</li>
	<li>STAR (for RNA-Seq)</li>
</ul>
<b>General statistics pertaining to reference assembly</b>
<ul class='a'>
	<li>Samtools</li>
	<li>Picard</li>
	<li>BBtools</li>
</ul>
<b>SNP and INDEL calling</b>
<ul class='a'>
	<li>Samtools/bcftools</li>
	<li>GATK</li>
	<li>iVAR</li>
</ul>
<b>Additional SNV filtering and annotation</b>
<p>In our case, we'll also be generating consensus genomes with iVAR</p>
</blockquote>
<br>



