<HTML><head><title>Bioinformatics Workshop</title></head><body>
<style>
#h3 {
 background: lightgray;
 white-space: nowrap;
}

table, th, td {
  border:1px solid black;
}

ul.a {
  list-style-type: circle;
}

pre.code {
        background-color: #EEE;
        padding: 5px 0px;
        margin-top: 5px;
        margin-bottom: 15px;
}

</style>

<h1>Tutorial 3 continued</h1>
<br>
<p> Now we will use our full fastq files downloaded from the SRA.</p>
<br>

<hr>
<p><b>Perform the same assembly with Shovill</b></p>
<hr>
<li>Use a Dockerized version of Shovill to assemble the full fastq files from the SRA.</li>
<blockquote>
<pre class="code">
docker run --rm -v $(pwd):/data -w /data staphb/shovill shovill --R1 SRR10971381_1.fastq --r2 SRR10971381_2.fastq --trim --outdir wuhan_assembly_shovill
</pre>
<p>As mentioned previously, Shovill is a faster implementation of SPAdes. Notice how much faster Shovill ran even with all of our data!
We are using the <code>--trim</code> option to remove adapters. Combined with the automatic read-error correction implemented, we can start with
our raw reads.  The one caveat is that Shovill doesn't have a metagenome mode.</p>
</blockquote>
<li>Briefly explore the Shovill assembly and its summary output to see how the program compares to SPAdes.</li>
<blockquote>
<p>The final contigs are in a file called 'contigs.fa' and are conveniently organized by descending length. Also notice that the naming conventions
for the contigs are much more manageable (although the original name assigned by SPAdes is also included).</p>
<pre style='font-family:Times New Roman'>
Did Shovill recover the Wuhan genome?
What assembler would you choose based on this exercise?</pre>
</blockquote>

<hr>
<p><b>Perform a metagenome assembly with Megahit.</p></b>
<hr>

<li>Follow the instructions on the GitHub page for <a style='color:blue' href="https://github.com/voutcn/megahit">Megahit</a> to obtain a pre-built binary.</li>
	<blockquote><p>Megahit should be installed in your home directory in a directory called 'MEGAHIT-1.2.9-Linux-x86_64-static.
	if you <code>cd</code> into this directory, you'll see another directory called 'bin.' This is a standard place for compiled software to be located.
	If you look in the 'bin' directory, you should see the Megahit executable.</p>
	<p>Remember, since this program is installed locally, you will have to provide the absolute path to the program when invoking it.</p>
	</blockquote>
<br>

<li>Use default parameters to assemble the metagenome with Megahit.</li>
<blockquote>
<pre class="code">
MEGAHIT-1.2.9-Linux-x86_64-static/bin/megahit -1 SRR10971381-trimmed_1.fastq -2 SRR10971381-trimmed_2.fastq -o wuhan_assembly_megahit
</pre>
<ol>
<li value='1'>
Generate the same summary statistics for your Megahit assembly as you did for your SPAdes assembly.</li>
<li>Did both assemblies recover the Wuhan-1 genome? </li>
<li>If so, are they similar lengths and coverage?</li>
<li>Are both identical to Wuhan-1?</li>
<li>Other tools to try in helping you evaluate the quality of your genomes are: <a style='color:blue' href="https://pangolin.cog-uk.io/">Pangolin</a>
(also available as a Docker container on your VM) and <a style='color:blue' href='https://clades.nextstrain.org/'>Nextclade</a>.
</ol>
</blockquote>
<br>

<hr>
<p><b>Compare the performance of SPAdes and Megahit with <a style='color:blue' href="https://github.com/ablab/quast">QUAST</a>.</p></b>
<hr>

<blockquote>
<p>Quast is a genome (and metagenome) evaluation tool that can compare the quality of assemblies produced by different programs.
It will output the number of large contigs for each assembly, the length of the longest one, the N50, and the number of genes predicted.
When a reference assembly is provided, Quast will output additional statistics, such as the number of misassemblies, the percentage of the reference genome
recovered, etc.</p>
</blockquote>


<br>
<br>
</body>
</HTML>